tests/agent_server/test_dryrun.py

```diff
@@ -313,3 +313,135 @@ async def test_dryrun_scale(resource_container, server, client, environment, age
+            "id": "test::Resource[agent1,key=key1],v=%d" % version,
+            "send_event": False,
+            "purged": False,
+            "requires": ["test::Resource[agent1,key=key2],v=%d" % version],
+        },
+        {
+            "key": "key2",
+            "value": "value2",
+            "id": "test::Resource[agent1,key=key2],v=%d" % version,
+            "send_event": False,
+            "requires": [],
+            "purged": False,
+        },
+        {
+            "key": "key3",
+            "value": None,
+            "id": "test::Resource[agent1,key=key3],v=%d" % version,
+            "send_event": False,
+            "requires": [],
+            "purged": True,
+        },
+        {
+            "key": "key4",
+            "value": execute.util.Unknown(source=None),
+            "id": "test::Resource[agent2,key=key4],v=%d" % version,
+            "send_event": False,
+            "requires": [],
+            "purged": False,
+        },
+        {
+            "key": "key5",
+            "value": "val",
+            "id": "test::Resource[agent2,key=key5],v=%d" % version,
+            "send_event": False,
+            "requires": ["test::Resource[agent2,key=key4],v=%d" % version],
+            "purged": False,
+        },
+        {
+            "key": "key6",
+            "value": "val",
+            "id": "test::Resource[agent2,key=key6],v=%d" % version,
+            "send_event": False,
+            "requires": ["test::Resource[agent2,key=key5],v=%d" % version],
+            "purged": False,
+        },
+    ]
+
+    status = {"test::Resource[agent2,key=key4]": const.ResourceState.undefined}
+    result = await client.put_version(
+        tid=environment,
+        version=version,
+        resources=resources,
+        resource_state=status,
+        unknowns=[],
+        version_info={},
+        compiler_version=get_compiler_version(),
+    )
+    assert result.code == 200
+
+    mod_db = await data.ConfigurationModel.get_version(uuid.UUID(environment), version)
+    undep = await mod_db.get_undeployable()
+    assert undep == ["test::Resource[agent2,key=key4]"]
+
+    undep = await mod_db.get_skipped_for_undeployable()
+    assert undep == ["test::Resource[agent2,key=key5]", "test::Resource[agent2,key=key6]"]
+
+    result = await client.dryrun_trigger(uuid.uuid4(), version)
+    assert result.code == 404
+    result = await client.dryrun_trigger(environment, 123456789)
+    assert result.code == 404
+
+    # request a dryrun with correct parameters
+    result = await client.dryrun_trigger(environment, version)
+    assert result.code == 200
+    dry_run_id = result.result["data"]
+
+    # get the dryrun results
+    result = await client.dryrun_list(environment, version)
+    assert result.code == 200
+    assert len(result.result["dryruns"]) == 1
+
+    while result.result["dryruns"][0]["todo"] > 0:
+        result = await client.dryrun_list(environment, version)
+        await asyncio.sleep(0.1)
```

Comment:
Please use `utils.retry_limited`, which includes a timeout. This code may hang the test suite indefinitely.